tpls/desul/include/desul/atomics/Compare_Exchange_OpenACC.hpp

```diff
@@ -0,0 +1,266 @@
+                  "operation in the OpenACC backend\n");
+  }
+  // Acquire a lock for the address
+  //while (!lock_address_openacc((void*)dest, scope)) {
+  //}
+  //device_atomic_thread_fence(MemoryOrderAcquire(), scope);
+  T return_val = *dest;
+  *dest = value;
+  //device_atomic_thread_fence(MemoryOrderRelease(), scope);
+  //unlock_address_openacc((void*)dest, scope);
+  return return_val;
+}
+
+#pragma acc routine seq
+template <class T, class MemoryOrder, class MemoryScope>
+std::enable_if_t<std::is_integral<T>::value && (sizeof(T) == 4), T> 
+	device_atomic_exchange(T* const dest, T value, MemoryOrder, MemoryScope) {
+  static_assert(sizeof(unsigned int) == 4,
+                "this function assumes an unsigned int is 32-bit");
+  unsigned int return_val = atomicExch(reinterpret_cast<unsigned int*>(dest),
+                                       reinterpret_cast<unsigned int&>(value));
+  return reinterpret_cast<T&>(return_val);
+}
+
+#pragma acc routine seq
+template <class T, class MemoryOrder, class MemoryScope>
+std::enable_if_t<std::is_integral<T>::value && (sizeof(T) == 8), T> 
+	device_atomic_exchange(T* const dest, T value, MemoryOrder, MemoryScope) {
+  static_assert(sizeof(unsigned long long int) == 8,
+                "this function assumes an unsigned long long int is 64-bit");
+  unsigned long long int return_val = atomicExch(reinterpret_cast<unsigned long long int*>(dest),
+                                       reinterpret_cast<unsigned long long int&>(value));
+  return reinterpret_cast<T&>(return_val);
+}
+
+#pragma acc routine seq
+template <class T, class MemoryOrder, class MemoryScope>
+std::enable_if_t<std::is_floating_point<T>::value && (sizeof(T) == 4), T> 
+	device_atomic_exchange(T* const dest, T value, MemoryOrder, MemoryScope) {
+  static_assert(sizeof(float) == 4,
+                "this function assumes an float is 32-bit");
+  float return_val = atomicExch(reinterpret_cast<float*>(dest),
+                                       reinterpret_cast<float&>(value));
+  return reinterpret_cast<T&>(return_val);
+}
+
+#ifndef DESUL_CUDA_ARCH_IS_PRE_PASCAL
+
+#pragma acc routine seq
+template <class T, class MemoryOrder, class MemoryScope>
+std::enable_if_t<std::is_floating_point<T>::value && (sizeof(T) == 8), T> 
+	device_atomic_exchange(T* const dest, T value, MemoryOrder, MemoryScope) {
+  static_assert(sizeof(double) == 8,
+                "this function assumes an double is 64-bit");
+  double return_val = atomicExch(reinterpret_cast<double*>(dest),
+                                       reinterpret_cast<double&>(value));
+  return reinterpret_cast<T&>(return_val);
+}
+
+#else
+
+#pragma acc routine seq
+template <class T, class MemoryOrder, class MemoryScope>
+std::enable_if_t<std::is_floating_point<T>::value && (sizeof(T) == 8), T> 
+	device_atomic_exchange(T* const dest, T value, MemoryOrder, MemoryScope scope) {
+  if (acc_on_device(acc_device_not_host)) {
```

Comment:
I understand that but the implementation will also not work correctly on the host (for host-parallel backends) so I would expect to always print an error message if you call this function (and if it's not called on the host, which would be preferable since we have other implementations for that anyway, there is also no point in checking).

Reply:
The current OpenACC atomic implementation supports the host only as a serial backend due to several unresolved design issues (e.g., distinguishing a serial host, OpenMP-parallel host, and OpenACC-parallel host).